### ISSUE TITLE
New rule: react-set-state-with-function

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint-react#80]
Enforce usage of the callback form of `this.setState`

```ts
// bad
this.setState({ ... });
// good
this.setState((prevState, props) => ...);
```

**Comment by vhfmag** [comment]
I don't know much about tslint and tsutils API, so that may be not feasible, but I think the ideal would be that the rule allows objects if props and state aren't used and enforce callback usage otherwise. Does that sound feasible?

**Comment by benbankes** [comment]
It seems like disallowing `this.setState({...});` might be a little too restrictive.  It seems like a better approach might be #172

**Comment by karfau** [comment]
From what I understand, the rule suggested in #172 is more flexible and covers more cases then this one. They have some overlap though and I can imagine a team deciding for this rule that might be simpler to implement and thus be more performent. So not sure if those two are duplicates or not.
Maybe the behavior suggested here could be provided by an option in the other rule to stop ading overlapping rules that might be confusing to choose from and configure.

**Comment by adidahiya** [comment]
closing in favor of #172
